### PR TITLE
Fixes #22072: Display compliance by directive

### PR DESCRIPTION
**Pull request by ElaadF**
https://issues.rudder.io/issues/22072
- [x] Compliance by directive Id
- [x] Compliance on each directive
- [x] Export CSV API + frontend

image

### CODE REVIEW

**Comment by ElaadF** [comment]
Commit modified

**Comment by ElaadF** [comment]
PR rebased

**Comment by ElaadF** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-web/src/main/elm/directivecompliance/sources/FakeUtils.elm [hunk 1]
```diff
@@ -0,0 +1,40 @@
+module FakeUtils exposing (..)
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-web/src/main/elm/directivecompliance/sources/FakeUtils.elm:1]: You should remove FakeUtils

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR rebased

**Comment by ElaadF** [comment]
PR rebased

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR rebased

**Comment by ElaadF** [comment]
PR updated with a new commit

**Comment by ElaadF** [comment]
PR rebased

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala [hunk 2]
```diff
@@ -167,6 +167,20 @@ object ComplianceApi       extends ApiModuleProvider[ComplianceApi] {
...
   }
 
+  final case object GetDirectivesCompliance extends ComplianceApi with ZeroParam with StartsAtVersion10 with SortIndex {
+    val z              = implicitly[Line].value
+    val description    = "Get all directive's compliance"
+    val (action, path) = GET / "compliance" / "directives"
+    val dataContainer  = Some("directivesCompliance")
+  }
+
+  final case object ExportDirectiveComplianceCSV extends ComplianceApi with OneParam with StartsAtVersion10 with SortIndex {
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala:177]: I would rather have the URL to be compliance/directives/id/csv

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala:177]: (in fact i would prefer compliance/directives/id?format=csv

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 3]
```diff
@@ -190,6 +193,73 @@ class ComplianceApi(
...
+      (for {
+        level     <- restExtractor.extractComplianceLevel(req.params)
+        t1         = System.currentTimeMillis
+        precision <- restExtractor.extractPercentPrecision(req.params)
+        id        <- DirectiveId.parse(directiveId).toBox
+        t2         = System.currentTimeMillis
+
+        directive <- complianceService.getDirectiveCompliance(id, level)
+        t3         = System.currentTimeMillis
+        _          = TimingDebugLogger.trace(s"API Export compliance to CSV - getting query param in ${t2 - t1} ms")
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:219]: Why do you put your traces at the end ? you shoud make them just below what they are doing

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 4]
```diff
@@ -239,6 +309,59 @@ class ComplianceApi(
...
+        precision   <- restExtractor.extractPercentPrecision(req.params)
+        t2           = System.currentTimeMillis
+        fullLibrary <- readDirective.getFullDirectiveLibrary().toBox ?~! "Could not fetch Directives"
+        t3           = System.currentTimeMillis
+        directiveIds = fullLibrary.allDirectives.values.filter(!_._2.isSystem).map(_._2.id).toList
+//        id        <- DirectiveId.parse(directiveId).toBox
+        t4           = System.currentTimeMillis
+        directives   = directiveIds.flatMap(complianceService.getDirectiveCompliance(_, level))
+        t5           = System.currentTimeMillis
+        _            = TimingDebugLogger.trace(s"API GetDirectives - getting query param in ${t2 - t1} ms")
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:337]: same here and it misses some cases

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 5]
```diff
@@ -239,6 +309,59 @@ class ComplianceApi(
...
+        if (version.value <= 6) {
+          directives.map(_.toJsonV6)
+        } else {
+          val json = directives.map(
+            _.toJson(
+              level.getOrElse(10),
+              precision.getOrElse(CompliancePrecision.Level2)
+            )
+          ) // by default, all details are displayed
+          val t4 = System.currentTimeMillis
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:350]: Strange that it compiles with two T4

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala (hunk 2 above)
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala:177]: Yes i would définitely prefer one endpoint and choose the format as a parameter

**Review by VinceMacBuche — APPROVED** [review]

**Comment by VinceMacBuche** [comment]
I'm merging but we should treat the feedbacks during beta 2

**Comment by ElaadF** [comment]
OK, squash merging this PR

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala (hunk 2 above)
Comment (ncharles) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala:177]: why does it starts at version 10 and not the latest version ?

File: webapp/sources/rudder/rudder-web/src/main/elm/directivecompliance/sources/ApiCalls.elm [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    req
+
+getNodesList : Model -> Cmd Msg
+getNodesList model =
+  let
+    req =
+      request
+        { method  = "GET"
+        , headers = []
+        , url     = getUrl model ["nodes"] []
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-web/src/main/elm/directivecompliance/sources/ApiCalls.elm:56]: i'm unsure as to why you need that. If it's really necessary, you could fetfh nly the minimal version of the node, or minimal+ exactly what you need

**Pull request merged**